[libpng] Tweak build rules for drake

Tweaks Linux intel rules and pnglibconf.h
--- BUILD.bazel
+++ BUILD.bazel
@@ -22,6 +22,14 @@ license(
     license_text = "LICENSE",
 )

+config_setting(
+    name = "build_intel",
+    constraint_values = [
+        "@platforms//cpu:x86_64",
+        "@platforms//os:linux",
+    ],
+)
+
 genrule(
     name = "pnglibconf",
     srcs = ["scripts/pnglibconf.h.prebuilt"],
@@ -65,7 +73,7 @@ cc_library(
             "arm/filter_neon_intrinsics.c",
             "arm/palette_neon_intrinsics.c",
         ],
-        "@platforms//cpu:x86_64": [
+        ":build_intel": [
             "intel/filter_sse2_intrinsics.c",
             "intel/intel_init.c",
         ],
@@ -83,6 +91,20 @@ cc_library(
         "png.h",
         "pngconf.h",
     ],
+    copts = [
+        "-w",
+        # Turn off <config.h> guessing. It should be implicitly off by default,
+        # but it would be a disaster if the default somehow didn't work.
+        "-DPNG_NO_CONFIG_H=1",
+        # Don't allow the pngpriv.h auto-sensing of Neon to take effect.
+        "-DPNG_ARM_NEON_OPT=0",
+        "-DPNG_ARM_NEON_IMPLEMENTATION=0",
+    ] + select({
+        ":build_intel": [
+            # Use SSE4.1 (or earlier) SIMD instructions.
+            "-msse4.1",
+        ],
+    }),
     includes = ["."],
     linkopts = select({
         "@platforms//os:windows": [],
@@ -95,8 +117,9 @@ cc_library(
         "@platforms//cpu:armv7": [
             "PNG_ARM_NEON_OPT=2",
         ],
-        "@platforms//cpu:x86_64": [
+        ":build_intel": [
             "PNG_INTEL_SSE_OPT=1",
+            "PNG_INTEL_SSE_IMPLEMENTATION=3",
         ],
         "@platforms//cpu:x86_32": [
             "PNG_INTEL_SSE_OPT=1",
new file mode 100644
--- scripts/pnglibconf.h.prebuilt
+++ scripts/pnglibconf.h.prebuilt
@@ -231,3 +231,12 @@
 #define PNG_sRGB_PROFILE_CHECKS 2
 /* end of settings */
 #endif /* PNGLCONF_H */
+#pragma once
+
+/* The following are Drake's adjustments to libpng. */
+
+/* Always use the macros, never the slow out-of-line functions. */
+#undef PNG_READ_INT_FUNCTIONS_SUPPORTED
+
+/* Suppress benign eXIf warnings, as they produce unnecessary test output. */
+#undef PNG_READ_eXIf_SUPPORTED
